### DISCUSSION TITLE
How to handle internal database tables that are too large？

### DISCUSSION THREAD

**Discussion by avidbyte** [rabbitmq/rabbitmq-server#5081] [Other] [answered]
version：RabbitMQ 3.8.8,Erlang 23.1,single node,Only the delay plugin is used.
`rabbitmqctl status`
mnesia: 9.0151 gb (79.89 %)
It was found through the web interface that internal database tables took up a lot of memory. What is the reason for this?

**Comment by avidbyte** [comment]
I don't have any message accumulation on my queue

**Comment by mkuratczyk** [comment, marked as answer]
Messages are not stored in Mnesia, metadata is.

1. Can you tell us how much entities, such as vhosts, queues, bindings, etc you have?

2. Can you share the output of:
```
rabbitmqctl eval 'mnesia:schema().' | grep -E 'Properties|size'
```
I hope this will tell us what is consuming so much space (what you shared is memory usage but I'd expect it to be correlated with the number of entries in the table).

**Reply by avidbyte** [reply]
There is only one vhosts, there are 5 queues, and the 5 queues are all delayed queues

**Reply by avidbyte** [reply]
root@myRabbit2:/# rabbitmqctl eval 'mnesia:schema().' | grep -E 'Properties|size'
-- Properties for gm_group table --- 
size                 -> 0
-- Properties for mirrored_sup_childspec table --- 
size                 -> 0
-- Properties for rabbit_delayed_messagerabbit@myRabbit2 table --- 
size                 -> 6008272
-- Properties for rabbit_delayed_messagerabbit@myRabbit2_index table --- 
size                 -> 1051188
-- Properties for rabbit_durable_exchange table --- 
size                 -> 12
-- Properties for rabbit_durable_queue table --- 
size                 -> 5
-- Properties for rabbit_durable_route table --- 
size                 -> 5
-- Properties for rabbit_exchange table --- 
size                 -> 12
-- Properties for rabbit_exchange_serial table --- 
size                 -> 0
-- Properties for rabbit_listener table --- 
size                 -> 4
-- Properties for rabbit_node_maintenance_states table --- 
size                 -> 1
-- Properties for rabbit_queue table --- 
size                 -> 5
-- Properties for rabbit_reverse_route table --- 
size                 -> 5
-- Properties for rabbit_route table --- 
size                 -> 5
-- Properties for rabbit_runtime_parameters table --- 
size                 -> 2
-- Properties for rabbit_semi_durable_route table --- 
size                 -> 5
-- Properties for rabbit_topic_permission table --- 
size                 -> 0
-- Properties for rabbit_topic_trie_binding table --- 
size                 -> 0
-- Properties for rabbit_topic_trie_edge table --- 
size                 -> 0
-- Properties for rabbit_topic_trie_node table --- 
size                 -> 0
-- Properties for rabbit_user table --- 
size                 -> 1
-- Properties for rabbit_user_permission table --- 
size                 -> 1
-- Properties for rabbit_vhost table --- 
size                 -> 1
-- Properties for schema table --- 
size                 -> 26
-- Properties for tracked_connection_on_node_rabbit@myRabbit2 table --- 
size                 -> 5
-- Properties for tracked_connection_per_vhost_on_node_rabbit@myRabbit2 table --- 
size                 -> 1

**Reply by avidbyte** [reply]
Could it be because the rabbitmq service has been restarted several times recently

**Reply by mkuratczyk** [reply]
Right, so you do have accumulated messages, over 6 million it seems. They are just delayed. Delayed messages are in fact an exception to what I said and are indeed kept in Mnesia. https://github.com/rabbitmq/rabbitmq-delayed-message-exchange#limitations mentions says:
```
Current design of this plugin doesn't really fit scenarios with a high number of delayed messages (e.g. 100s of thousands or millions). See #72 for details.
```

**Reply by avidbyte** [reply]
i see， thank you very mach

**Comment by michaelklishin** [comment]
Most likely your applications leak queues, bindings, or potentially users or permissions.

FYI, RabbitMQ 3.8 goes [completely out of support](https://www.rabbitmq.com/versions.html) next month.
